File: 08/solution.js
import {check, readInput} from "../Utils.js";

function part1(input) {
    let totalSize = 0
    let totalInMemorySize = 0
    for (const line of input) {
        totalSize += line.length
        for (let i = 1; i < line.length - 1; i++) {
            if(line[i] === '\\') {
                if(line[i+1] === 'x') { i += 3}
                else { i++ }
            }
            totalInMemorySize++
        }
    }
    return totalSize - totalInMemorySize
}

function part2(input) {
    let totalSize = 0
    let totalEncodedSize = 0
    for (const line of input) {
        totalSize += line.length
        for (let i = 0; i < line.length; i++) {
            if(line[i] === '\\' || line[i] === '\"') {
                totalEncodedSize++
            }
        }
        totalEncodedSize += line.length + 2
    }
    return totalEncodedSize - totalSize
}

const testInput = readInput("test_input.txt")
const input = readInput("input.txt")

check(part1(testInput), 12)
console.log("Part 1: " + part1(input))

check(part2(testInput), 19)
console.log("Part 2: " + part2(input))

File: 08/solution.rb
require_relative "../utils"
include Utils

def part1(input)
  registers = {}
  registers.default = 0
  input.each do |line|
    register, op, value, _, cond_register, cond, cond_value = line.split
    if eval("#{registers[cond_register]} #{cond} #{cond_value}")
      sign = 1
      if op == "dec"
        sign = -1
      end
      registers[register] += value.to_i * sign
    end
  end
  registers.values.max
end

def part2(input)
  registers = {}
  registers.default = 0
  max_value = 0
  input.each do |line|
    register, op, value, _, cond_register, cond, cond_value = line.split
    if eval("#{registers[cond_register]} #{cond} #{cond_value}")
      sign = 1
      if op == "dec"
        sign = -1
      end
      registers[register] += value.to_i * sign
      max_value = [registers.values.max, max_value].max
    end
  end
  max_value
end

input = read_input("input")
test_input = read_input("test_input")

check(1, part1(test_input))
puts "Part 1: #{part1(input)}"

check(10, part2(test_input))
puts "Part 2: #{part2(input)}"